Faster search, custom exclude list, and more relaxed mode exclusions

joanbm opened the pull request:
Hello,

This pull request includes quite a few changes, so feel free to reject it if you only want to accept a part of it or simply want it more atomized if it eases the review process. Maybe consider it more like a RFC.

The changes are the following:

* **Faster search:** I noticed that the program is often CPU bound on SSDs, and the bottleneck is that when specifying many paths with `-wholename`, `find (findutils)` does not seem to be able to optimize the query, but rather just does a plain linear search of all strings over any path.

After some tests, I noticed that using the `-regex` option (with a regular expression equivalent to the original query) is much faster. I haven't looked at the internals of this, but my guess is that `find (findutils)` uses an optimized regex library when evaluating a regular expression. 

The script can now run in relaxed mode in about 1 second on my system, while it takes a few seconds with the non-regex version.

Technically, in order to avoid hardcoding a massive incomprehensible regexp, I put the list of paths to exclude in a bash array, and then use a function to turn it into a regular expression. The paths are written as plaintext, but the * character can be used as a wildcard, like in glob. 

(Note: The `node_modules` folder, which has a different filter than the other paths in the script, is still handled manually in a hacky way for now).

* **Custom exclude list:** The user can create a file called `/etc/lostfiles.excludes` to specify files that he wishes to exclude in relaxed mode. This is very useful for excluding custom configuration files, binaries, etc.

The format is also a list of plaintext paths, and * can be used as a wildcard.

```
/etc/fonts/conf.d/99-zealcharm-default-fonts.conf
/etc/netctl/zealcharm-network-keepalive.service
/etc/NetworkManager/conf.d/zealcharm-dhcp-client.conf
/etc/pam.d/*
[...]
```

Technically this builds on the previous feature, adding the contents of the `/etc/lostfiles.excludes` file to the list of exclude paths before building the regexp.

This is related to #35. I saw that there's an open PR (#41) that is also about this, albeit with a more powerful solution that allows any filter of find instead of just a list of plaintext paths with optional asterisk. For now I have not integrated this PR with it.

**More relaxed exclusions:** Simply more autogenerated configuration files, etc. have been added to the relaxed excludes list, those are some from issue #36 plus a few more I found on my system. I added a comment after the path specifying where it comes from.

ArchangeGabriel commented on common/lostfiles.in:
```diff
@@ -39,127 +35,197 @@ while getopts 'sz' OPTION; do
+  )
+  
+  exclude_list_regex=$(miniglob_list_to_regex "${exclude_list[@]}")
+  echo -n "( -regex \"|$exclude_list_regex|\" -prune )"
```
Please note that this is not how the strict mode works since #40. Using an exclude list is not a good idea for it, and I believe this can only be slower than the current way of doing things.

ArchangeGabriel commented on common/lostfiles.in:
```diff
@@ -39,127 +35,197 @@ while getopts 'sz' OPTION; do
+    "/var/spool"
+    "/var/db/sudo"
+    "/var/tmp"
+  )
```
Same here, a lot of those files do not have to be excluded in the first place if you do not search for paths outside the ones I’ve listed before.

ArchangeGabriel commented:
This looks interesting generally speaking, but I think we can do better than that.

I would remove any excludes hardcoding from the script, and instead ship the list as a default `lostfiles.excludes` configuration file (which is partly what #41 does btw), which users can then modify to suit their needs (by adding/removing paths to/from the list).

Now I’m wondering if we need as much flexibility as provided by #41, or if having the strict mode check all listed path and the relaxed mode ignoring the `lostfiles.excludes` on top of that would be enough. Do any of you have use cases for more than that? I don’t see any reasons to look outside of where packages are allowed to have files, and I don’t see what would be the purpose of having strict + relaxed exclude lists.

ArchangeGabriel commented:
Also tagging @blacktemplar to be sure to have everyone in this discussion.

joanbm commented:
Thank you for the review, I will think about the proposed changes and update the PR soon.

joanbm commented:
I like the idea of moving the hard-coded parameters in the script to configuration files which can then be customized by the user, as was implemented in @blacktemplar 's fork.

However, I made some changes, and I have updated the PR with a new proposal for the configuration files, with some ideas from the comments, some from @blacktemplar 's fork, and some of my own.

* I hardly ever use the strict mode and I think adding excludes for the strict mode kind of defeats its purpose, so for simplicity's sake I didn't make the strict mode configurable like in @blacktemplar 's fork, so it's still a (very short) hardcoded list of paths to include.

* On the other hard, I like being flexible about configuration possibilities in relaxed mode (even include paths, for example I prefer the script to look at all folders under `/` to ensure I don't have leftovers which could come from tests, typos, buggy scripts, misbehaving AUR packages, etc.)

  In my old PR, I added a single configuration file that only contained user-defined exclude paths, while the hardcoded include paths, exclude paths and custom exclude filters (the filter of node_modules/npm and custom kernel modules) remained.

  On the other hard, in @blacktemplar 's fork, the include paths and exclude paths were moved to two separate configuration files, which can then be customized by the user. However, one of the custom exclude filters was still hardcoded and the other was tacked along with the include paths.

  With my new proposal I attempt to combine some ideas from both, and additionally completely removing the hardcoded filters in the script. This is achieved through a single configuration file, that defines the include paths, exclude paths and custom exclude filters. Each kind defined by a prefix at the start of a line:

  * Lines beginning with a '+' are include paths, e.g.  `+/boot`
  * Lines beginning with a '-' are exclude paths, e.g. ` -/boot/initramfs*`. (Asterisks are allowed as wildcards)
  * Lines beginning with a '$' are custom filters (actually parameters passed as-is to `find`), e.g. ` $\( -not \( -path '/usr/lib/modules/'``uname -r`` -prune \) \) `

What do you all think about it? I think the configuration format is pretty each to understand and the code is not really long or complicated, but don't hold your comments about overengineering if you think so 😅.

joanbm commented:
Any news?

graysky2 commented:
@joanbm - Please rebase to a single commit for review.

joanbm commented:
Sorry for the delay - It should all be in one commit now.

ArchangeGabriel commented:
@graysky2 Ping.

astewartau commented:
Just commenting that I would like to see this functionality! :+1:

ArchangeGabriel commented:
@graysky2 Friendly ping. ;)

graysky2 commented:
I'm thinking we would want some config file error check function if this get merged.  Users an introduce errors here that would break the script.

joanbm commented:
Right now lines in `lostfiles.excludes` which don't start with `+-$` are ignored, I guess we should define some prefix (e.g. #) as comment and fail otherwise. From a quick look, I think there's nothing else in the configuration file that can be wrong.

It should also be possible to accept any shell pattern as excludes (without resorting to the `$` wildcard), by taking the code that comes from the similar PR from aconfmgr linked a few coments above ( https://github.com/CyberShadow/aconfmgr/pull/74 ).

joanbm commented:
I've updated the PR to add some basic validation for the configuration file (check that lines starts with either +, -, $ or #). I've also fixed some minor problems the previous PR had (fix shellcheck warnings, better commit messages, explain how comment lines work in the manpage).

I guess there's some further possibilities for checking for misuse of the configuration file (e.g. no duplicated include/exclude filters) but they don't add much value IMO. I can take a look if you want though.

Since the squashed diff is pretty big, you can check it split in 4 commits in this branch just in case: https://github.com/joanbm/lostfiles/tree/rebase_pr42

PS: I didn't realize I was generating this GitHub spam it while working on a separate branch, I'd have avoided it otherwise

graysky2 commented:
I just merged into my dev branch.  Thanks for the contribution.  Not sure if I did it the github way because the PR didn't automatically close.

https://github.com/graysky2/lostfiles/commit/bddf4f4eee760c926da300baba79c9cca6f3a140
